### PR TITLE
Complete response

### PR DESCRIPTION
**Pull request by Del-S**
- Added a wrapping object for complete response which contains Result and retrofit2 + okhttp3 responses.

Depends on: https://github.com/eManPrague/kaal/pull/34

### CODE REVIEW

File: kaal-infrastructure/src/main/kotlin/cz/eman/kaal/infrastructure/api/retrofit/KaalRetrofitCaller.kt [hunk 1]
```diff
@@ -2,127 +2,210 @@ package cz.eman.kaal.infrastructure.api.retrofit
...
-    suspend fun <Dto, T> callResultResponseOptional(
+    suspend fun <Dto, T> callResultCompleteOptional(
         responseCall: suspend () -> Response<Dto>,
         errorMessage: () -> String?,
         map: suspend (Dto) -> T
-    ): Pair<Result<Optional<T>>, Response<Dto>?>
+    ): CompleteRetrofitResponse<Optional<T>, Dto> {
+        val response = callResultCompleteNull(responseCall, errorMessage, map)
+        val mappedResult = response.result.map { data ->
+            if (data != null) {
```
Comment (LukasKrystekEman) [inline kaal-infrastructure/src/main/kotlin/cz/eman/kaal/infrastructure/api/retrofit/KaalRetrofitCaller.kt:147]: the same construct has been used for response of callResultNull in CallResultUnit method, i recommend to extract this mapper to separated method

File: kaal-infrastructure/src/main/kotlin/cz/eman/kaal/infrastructure/api/retrofit/KaalRetrofitCaller.kt [hunk 2]
```diff
@@ -2,127 +2,210 @@ package cz.eman.kaal.infrastructure.api.retrofit
...
-     * @since 0.9.0
-     */
-    suspend fun <Dto> callResultUnit(
-        responseCall: suspend () -> Response<Dto>,
-        errorMessage: () -> String?,
-        map: suspend (Dto) -> Unit = { }
-    ): Result<Unit> = callResultNull(responseCall, errorMessage, map).map { }
+    ): Result<Optional<T>> {
+        return callResultNull(responseCall, errorMessage, map).map { data ->
+            if (data != null) {
```
Comment (LukasKrystekEman) [inline kaal-infrastructure/src/main/kotlin/cz/eman/kaal/infrastructure/api/retrofit/KaalRetrofitCaller.kt:85]: I recommend to use Elvis operator

File: kaal-infrastructure/src/main/kotlin/cz/eman/kaal/infrastructure/api/retrofit/KaalRetrofitCaller.kt [hunk 3]
```diff
@@ -140,4 +223,60 @@ interface KaalRetrofitCaller {
...
+     * @param map used to map [Dto] object to [T]
+     * @return @return [Result] with type [T] or null
+     * @see createApiErrorResult
+     * @since 0.9.0
+     */
+    private suspend fun <Dto, T> handleResponse(response: Response<Dto>, map: suspend (Dto) -> T): Result<T?> {
+        return when {
+            response.isSuccessful -> {
+                val body = response.body()
+                if (body != null) {
```
Comment (LukasKrystekEman) [inline kaal-infrastructure/src/main/kotlin/cz/eman/kaal/infrastructure/api/retrofit/KaalRetrofitCaller.kt:267]: isn't better to use Elvis?
```
body?.let {
   Result.success(map(body))
} ?:
 Result.success(null)
```

or better
`Result.success(body?.let{ map(body) })`

File: kaal-infrastructure/src/main/kotlin/cz/eman/kaal/infrastructure/api/retrofit/CompleteRetrofitResponse.kt [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+ * 2) Retrofit2 [response] with mapped [Dto] object and additional information.
+ * 3) Okhttp3 [httpResponse] with raw call data,
+ *
+ * @author eMan a.s.
+ * @since 0.9.0
+ */
+data class CompleteRetrofitResponse<T, Dto>(
+    val result: Result<T>,
+    val response: Response<Dto>?,
+    val httpResponse: okhttp3.Response?
```
Comment (smidf) [inline kaal-infrastructure/src/main/kotlin/cz/eman/kaal/infrastructure/api/retrofit/CompleteRetrofitResponse.kt:18]: Why to hold them as special property, if can be get via `Response.raw`?

**Review by smidf — APPROVED** [review]

**Pull request merged**